compiler/sem/varpartitions.nim

```diff
@@ -362,6 +362,40 @@ proc pathExpr(node: PNode; owner: PSym): PNode =
           break
       else:
         break
+    of nkElifExpr, nkElifBranch, nkElseExpr, nkElse, nkExceptBranch, nkOfBranch:
+      n = n.lastSon
+    of nkIfExpr, nkIfStmt, nkTryStmt, nkCaseStmt:
+      block samePathCheck:
+        let firstExpr =
+          case n.kind
+          of nkTryStmt:
+            n[0] # First element is try body
+          of nkCaseStmt:
+            n[1].lastSon # First branch is the second element
+          else:
+            n[0].lastSon
+        let firstValidPath = pathExpr(firstExpr, owner)
+        # The first expression should be a valid path
+        if firstValidPath == nil:
+          break samePathCheck
+
+        let remainderStart = if n.kind == nkCaseStmt: 2 else: 1
+        for idx in remainderStart ..< n.len:
+          # Only check branches with a type, as no types implies noreturn for
+          # expressions
```

Comment:
They're either noreturn or unit, the difference being that the latter doesn't "abort" the current expression like the former does. You might need to rely on the `endsInNoReturn` analysis.